velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -135,79 +127,101 @@ struct RegexpReplaceFunction {
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    callAscii(result, stringInput, pattern, replace, 1);
+  }
 
-    if (position > stringInput.size() + 1) {
-      result.resize(inputStringPiece.size());
-      std::memcpy(
-          result.data(), inputStringPiece.data(), inputStringPiece.size());
+  void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    if (performChecks(result, stringInput, pattern, replace, position - 1)) {
       return;
     }
+    performReplace(result, stringInput, pattern, replace, position - 1);
+  }
 
-    // Adjust the position for UTF-8 by counting the code points.
-    size_t utf8Position = 0;
-    size_t numCodePoints = 0;
-    while (numCodePoints < position - 1 && utf8Position <= stringInput.size()) {
-      int charLength =
-          utf8proc_char_length(inputStringPiece.data() + utf8Position);
-      VELOX_USER_CHECK_GT(
-          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
-      ++numCodePoints;
-      utf8Position += charLength;
+ private:
+  bool performChecks(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_USER_CHECK_GE(
+        position + 1, 1, "regexp_replace requires a position >= 1");
+    if (position > stringInput.size()) {
+      result = stringInput;
+      return true;
     }
-    if (utf8Position > stringInput.size() + 1) {
-      result.resize(inputStringPiece.size());
-      std::memcpy(
-          result.data(), inputStringPiece.data(), inputStringPiece.size());
-      return;
+
+    if (stringInput.size() == 0) {
+      if (pattern.size() == 0 && position == 1) {
+        result = replace;
+        return true;
+      }
+      if (pattern.size() > 0) {
+        result = stringInput;
+        return true;
+      }
     }
+    return false;
+  }
 
-    re2::StringPiece prefix(inputStringPiece.data(), utf8Position);
-    re2::StringPiece targetStringPiece(
-        inputStringPiece.data() + utf8Position,
-        inputStringPiece.size() - utf8Position);
+  void performReplace(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    re2::RE2* patternRegex = getRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
 
+    std::string prefix(stringInput.data(), position);
     std::string targetString(
-        targetStringPiece.data(), targetStringPiece.size());
-    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+        stringInput.data() + position, stringInput.size() - position);
```

Comment:
nit: I am not sure about this, but is stringInput.size() size in bytes or unicode code points ? (The arithmetic to compute the targetString might not be correct in former case).

Reply:
In this case, the position is dependent on the function that calls performReplace.

For example, in callAscii the position is always just the user passed position translated to a 0-indexed value:

```cpp
  void callAscii(
      out_type<Varchar>& result,
      const arg_type<Varchar>& stringInput,
      const arg_type<Varchar>& pattern,
      const arg_type<Varchar>& replace) {
    callAscii(result, stringInput, pattern, replace, 1);
  }

  void callAscii(
      out_type<Varchar>& result,
      const arg_type<Varchar>& stringInput,
      const arg_type<Varchar>& pattern,
      const arg_type<Varchar>& replace,
      const arg_type<int64_t>& position) {
    if (performChecks(result, stringInput, pattern, replace, position - 1)) {
      return;
    }
    performReplace(result, stringInput, pattern, replace, position - 1);
  }
```
In this case a character is always a single byte.

In the UTF-8 case, we convert the passed position to bytes in this block:

```cpp
    size_t start = functions::stringImpl::cappedByteLength<false>(
        stringInput, position - 1);
    if (start > stringInput.size() + 1) {
      result = stringInput;
      return;
    }
    performReplace(result, stringInput, pattern, replace, start);
```